Essentials/src/com/earth2me/essentials/commands/Commandsetwarp.java

```diff
@@ -10,35 +10,46 @@
 
 import static com.earth2me.essentials.I18n.tl;
 
-
 public class Commandsetwarp extends EssentialsCommand {
     public Commandsetwarp() {
         super("setwarp");
     }
 
     @Override
-    public void run(final Server server, final User user, final String commandLabel, final String[] args) throws Exception {
+    public void run(final Server server, final User user, final String commandLabel, final String[] args)
+            throws Exception {
         if (args.length == 0) {
             throw new NotEnoughArgumentsException();
         }
 
-        if (NumberUtil.isInt(args[0]) || args[0].isEmpty()) {
+        final String warpName = args[0];
+        final String safeWarpName = StringUtil.safeString(warpName);
+
+        if (NumberUtil.isInt(warpName) || warpName.isEmpty()) {
             throw new Exception(tl("invalidWarpName"));
         }
 
         final IWarps warps = ess.getWarps();
         Location warpLoc = null;
 
         try {
-            warpLoc = warps.getWarp(args[0]);
+            warpLoc = warps.getWarp(warpName);
         } catch (WarpNotFoundException | InvalidWorldException ignored) {
         }
 
-        if (warpLoc == null || user.isAuthorized("essentials.warp.overwrite." + StringUtil.safeString(args[0]))) {
-            warps.setWarp(user, args[0], user.getLocation());
-        } else {
+        final boolean warpExists = warpLoc != null;
+        final boolean canOverwrite = user.isAuthorized("essentials.warp.overwrite." + safeWarpName);
+        final boolean canCreate = user.isAuthorized("essentials.warp.set." + safeWarpName);
```

Comment:
Is this new permission backwards compatible with users on old versions of Essentials prior to this PR? Players with the permission `essentials.setwarp` should be able to create all warps by default. Permission defaults should be defined in `plugin.yml`.

Reply:
Ugh, guess not. Gotta fix

Reply:
@pop4959 should we keep `essentials.setwarp` as a fallback permission? I don't see any other ways since it's auto checked by Essentials' command handler. We could grant creating any warps via this permission and allow only specific if user has any *additional* perms like `essentials.warp.set.test`.
`essentials.setwarp` -> allow creating any warp
`essentials.setwarp` + `essentials.warp.set.test` -> allow creating only 'test' warp
Like that?

Reply:
How legacy support has been done for other permissions is normally that the base permission node acts as if the user has permission for all sub-nodes, and in this way there is no breakage with existing permissions.

E.g.
`essentials.setwarp=true` -> allow creating any warp
`essentials.setwarp=undefined/false` + `essentials.setwarp.test=true` -> allow creating only `test` warp

Reply:
I would like to add that the above is how I would probably try to go about implementing this, however if any other team members want to give another opinion on this that would be appreciated.

Reply:
@pop4959 @md678685 what do you guys think about reusing `per-warp-permission` setting? It's currently used for tp'ing to warps only, we could extent it to warp setting stuff or do we need a new option?

Reply:
It would change the existing behavior, and when it's regarding permissions I would like if you kept it unaffected and added a new option instead

Reply:
That sounds like a great idea, and if it were just me I would recommend that option. I don't believe minor breaking changes should be a blocking issue when it comes to improving software, but the current norm for Essentials is to make things absolutely idiot proof so that people who don't change their config / permissions *at all* still functions the same after an update. You'll have to wait for @md678685 for a final decision on that since he is the team lead.

Reply:
If it weren't for the fact that the behaviour of that config option was well-documented and well-known, I'd agree with pop. However, since this is an entirely separate feature, lumping it in with the existing config option will confuse server owners ("why do I need extra permissions to set warps that I didn't need before?"). I'd recommend putting this under a separate config option (`per-warp-edit-permissions`) that enables per-warp permissions for `setwarp` and `delwarp`.

Reply:
I agree with MD that the config options should remain separate.